### ISSUE TITLE
vmagent does not scrape metrics when the option -promscrape.streamParse is enabled

### ISSUE THREAD

**Issue by Sinketsu** [VictoriaMetrics/VictoriaMetrics#1017]
**Describe the bug**
After enabling `-promscrape.streamParse` option on vmagent (v1.52.0) we see some errors like 
```
error        VictoriaMetrics/lib/promscrape/scrapework.go:235        error when scraping "http://<hostname>/stats/prometheus" from job "<jobname>" with labels {instance="<hostname>",job="<jobname>"}: cannot read Prometheus exposition data: context deadline exceeded
```

And metrics are not collected from specific services from Consul.

**Version**
```
$ vmagent --version
vmagent-20210113-110050-tags-v1.52.0-0-g7ce87ebcb
```

**Comment by f41gh7** [comment]
As workaround, try to increase `scrape_timeout` for these targets.

 Need to investigate it.

**Comment by f41gh7** [comment]
@Sinketsu  increasing `scrape_timeout` is the only one solution for now. At previous releases scraping errors with `promscrape.streamParse` were suppressed, this commit https://github.com/VictoriaMetrics/VictoriaMetrics/commit/7976c22797bacf90845b22f972de61a4b41da38b fixed it. You can check metric `vm_promscrape_scrapes_failed_total` and `up` for those targets at previous release.

**Comment by valyala** [comment]
Note that `scrape_timeout` can be configured per specific job in Prometheus configs - see https://prometheus.io/docs/prometheus/latest/configuration/configuration/#scrape_config

**Comment by valyala** [comment]
I'm unsure this is a bug - vmagent can spend non-trivial amounts of time when scraping targets with millions of metrics. If the time required for processing all these metrics per each scrape exceeds `scrape_timeout` (it is set to [10 seconds](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/lib/promscrape/config.go#L872) by default), then the `cannot read Prometheus exposition data: context deadline exceeded` error is logged. This error was suppressed in vmagent releases prior v1.52.0, so it wasn't visible. Substituting `bug` label with `question` label for this issue then.

**Comment by Sinketsu** [comment]
When I disable the `-promscrape.streamParse` option, all metrics are collected from all services. When enable - some services failed and metrics are not collected.
Is it true that with the option enabled, vmagent spends more time for scrape?

**Comment by wf1nder** [comment]
In addition to the @Sinketsu comment: the service itself is responding very fast:

```
# time curl -s 'http://<metrics_endpoint>/stats/prometheus' > /dev/null

real	0m0.306s
user	0m0.012s
sys	0m0.012s
```

But it is possible that it returns slightly more metrics data than other services.
And yes, when streamParse is disabled, everything working fine, we not getting errors `context deadline exceeded`.

**Comment by valyala** [comment]
> Is it true that with the option enabled, vmagent spends more time for scrape?

Stream parsing may require slightly more CPU time compared to default parsing.

Default parsing is performed in the following steps:
1) Read full response body from target into vmagent memory
2) Parse the response body in memory.
3) Send the parsed data to the configured remote storage.

The `scrape_timeout` is checked only during the first step. The second and the third steps can take more than `scrape_timeout` without any errors.

Stream parsing is performed in the following steps:
1) Read a chunk of response body from target
2) Parse the read chunk
3) Send the parsed chunk to the configured remote storage.
4) Go to step 1 if response body has more data.

The `scrape_timeout` is checked during the first step. But it includes all the time required for performing steps 2-4 in the loop. That's why the `scrape_timeout` can be exceeded due to slow parsing or slow data writing to remote storage.

**Comment by wf1nder** [comment]
Thank you, and you were right about ```scrape_timeout``` option: when stream mode is disabled, the Scrape Duration for those endpoints indicates < 1 second. When I enabled stream mode, Scrape Duration increased to 11-12 seconds. So increasing ```scrape_timeout``` option to 20s for endpoint fixing scrape issues.

But, I think, this behavior is not obvious. "Scrape timeout" implied exactly as scrape timeout: the time it takes to get a response from endpoint, without taking into account the time for data processing endpoint's data inside vmagent.

Are there any plans to change the method of counting scrape duration in streamed mode?

**Comment by valyala** [comment]
@wf1nder , could you build `vmagent` from the commit 908e35affd31975bf808effa7072f59197d8d507 and verify whether it properly applies `scrape_timeout` for `stream_parse: true` targets? See [build instructions for vmagent](https://docs.victoriametrics.com/vmagent.html#how-to-build-from-sources).

**Comment by wf1nder** [comment]
Yes, it looks like everything is correct now. After upgrading to this version and enabling streamed mode, the number of targets in the "up" state did not decrease, as it was during the last attempt to enable the streamed mode (because some of them started timeouting).

Thank you!

PS. Reducing memory consumption by 5 times, and reducing CPU consumption by 2 times in the streamed mode - is impressive!

**Comment by valyala** [comment]
The bugfix for proper `scrape_timeout` usage when `stream_parse: true` is set has been included in [vmagent v1.59.0](https://github.com/VictoriaMetrics/VictoriaMetrics/releases). Closing this issue as fixed.

**Comment by valyala** [comment]
FYI, the timeout for [stream parsing mode](https://docs.victoriametrics.com/vmagent.html#stream-parsing-mode) is synced with the timeout for non-stream parsing mode after the 6e8611f301a72a67a5c4032da7e54cccfa1202bb . See https://github.com/VictoriaMetrics/VictoriaMetrics/issues/4847 for details.
